### PR TITLE
Cache tool

### PR DESCRIPTION
**Pull request by persiaAziz-zz**
Don't merge this please. I have a lot of debug prints here

### CODE REVIEW

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 1]
```diff
@@ -48,7 +54,12 @@ using ts::Errata;
 using ts::FilePath;
 using ts::MemView;
 using ts::CacheDirEntry;
-
+#define VOL_HASH_TABLE_SIZE 32707
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: No `#define` unless absolutely necessary. `constexpr`.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 2]
```diff
@@ -48,7 +54,12 @@ using ts::Errata;
...
+#define VOL_HASH_ALLOC_SIZE (8 * 1024 * 1024) // one chance per this unit
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: This should be a Scalar instance.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 3]
```diff
@@ -48,7 +54,12 @@ using ts::Errata;
...
+#define VOL_HASH_EMPTY 0xFFFF
+#define STORE_BLOCK_SHIFT 13
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: Not needed, handled by Scalar.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 4]
```diff
@@ -48,7 +54,12 @@ using ts::Errata;
...
+#define STORE_BLOCK_SIZE 8192
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: Not needed, handled by Scalar.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 5]
```diff
@@ -64,7 +75,9 @@ ts::CommandTable Commands;
 int OPEN_RW_FLAG = O_RDONLY;
 
 struct Stripe;
-
+std::vector<Stripe *> globalVec_stripe;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: This should be a member of `Cache`,not a global.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 6]
```diff
@@ -64,7 +75,9 @@ ts::CommandTable Commands;
...
+std::unordered_set<std::string> URLset;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: This shouldn't be a global, it should be embedded in the command function.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 7]
```diff
@@ -64,7 +75,9 @@ ts::CommandTable Commands;
...
+unsigned short *stripes_hash_table;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: This should be a member, not a global.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 8]
```diff
@@ -154,6 +168,7 @@ struct Stripe {
 
   int64_t _buckets;  ///< Number of buckets per segment.
   int64_t _segments; ///< Number of segments.
+  const char *hashText = nullptr;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:168]: This should be a `std::string`.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 9]
```diff
@@ -182,6 +197,16 @@ Stripe::Chunk::clear()
 
 Stripe::Stripe(Span *span, Bytes start, CacheStoreBlocks len) : _span(span), _start(start), _len(len)
 {
+  const char *diskPath        = span->_path.path();
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:197]: What is this used for, that `_span->_path.path()` can't be?

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 10]
```diff
@@ -182,6 +197,16 @@ Stripe::Chunk::clear()
...
+  const size_t hash_seed_size = strlen(diskPath);
+  const size_t hash_text_size = hash_seed_size + 32;
+  char *hash_text             = static_cast<char *>(ats_malloc(hash_text_size));
+  strncpy(hash_text, diskPath, hash_text_size);
+  snprintf(hash_text + hash_seed_size, (hash_text_size - hash_seed_size), " %" PRIu64 ":%" PRIu64 "", (uint64_t)_start,
+           (uint64_t)_len.count() * STORE_BLOCK_SIZE);
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: This should just be `_len`. If not, we'll need to update the version of Scalar here.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 11]
```diff
@@ -182,6 +197,16 @@ Stripe::Chunk::clear()
 
 Stripe::Stripe(Span *span, Bytes start, CacheStoreBlocks len) : _span(span), _start(start), _len(len)
 {
+  const char *diskPath        = span->_path.path();
+  const size_t hash_seed_size = strlen(diskPath);
+  const size_t hash_text_size = hash_seed_size + 32;
+  char *hash_text             = static_cast<char *>(ats_malloc(hash_text_size));
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:200]: Should be `std::string`.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 12]
```diff
@@ -497,6 +522,7 @@ Errata
 Cache::allocStripe(Span *span, int vol_idx, CacheStripeBlocks len)
 {
   auto rv = span->allocStripe(vol_idx, len);
+  std::cout << span->_path << ":" << vol_idx << std::endl;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:527]: Is this just temporary debugging?

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 13]
```diff
@@ -788,6 +815,34 @@ Cache::loadSpanConfig(FilePath const &path)
   return zret;
 }
 
+Errata
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:820]: Why load them in a table?

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 14]
```diff
@@ -1064,6 +1119,125 @@ Span::clearPermanently()
...
+  uint64_t used                = 0;
+  // estimate allocation
+  for (auto &elt : globalVec_stripe) {
+    printf("stripe length %" PRId64 "\n", elt->_len.count() * STORE_BLOCK_SIZE);
+    rtable_entries[i] = elt->_len.count() * STORE_BLOCK_SIZE / VOL_HASH_ALLOC_SIZE;
+    rtable_size += rtable_entries[i];
+    uint64_t x = elt->hash_id.fold();
+    // seed random number generator
+    rnd[i] = (unsigned int)x;
+    total += ((elt->_len.count() * STORE_BLOCK_SIZE) >> STORE_BLOCK_SHIFT);
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: Isn't this a no-op?

Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheTool.cc]: yah.. I calculated this to verify with the ATS implementation

File: cmd/traffic_cache_tool/CacheTool.cc (hunk 12 above)
Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheTool.cc:527]: yes

File: cmd/traffic_cache_tool/CacheTool.cc (hunk 9 above)
Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheTool.cc:197]: No use of this. I think I was looking for a shorter variable name

File: cmd/traffic_cache_tool/CacheTool.cc (hunk 6 above)
Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheTool.cc]: ok

**Pull request merged**